### ISSUE TITLE
How we represent qualifiers in data model?

### ISSUE THREAD

**Issue by skalee** [glossarist/iev-data#94]
From https://en.wiktionary.org/wiki/qualifier#English:

> (grammar) A word or phrase, such as an adjective or adverb, that describes or characterizes another word or phrase, such as a noun or verb; a modifier; that adds or subtracts attributes to another.

Can we consider it as another part of speech (although in fact it's not)? Or maybe we need a separate field for that?

Qualifier examples:

- [131-14-21](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=131-14-21)
- [131-14-23](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=131-14-23)
- [151-15-01](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=151-15-01)
- [151-15-02](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=151-15-02)
- [713-10-41](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=713-10-41)
- [826-14-15](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=826-14-15)

Here is how they are displayed in the Electropedia:

> ![Zrzut ekranu 2021-01-12 o 19 38 35](https://user-images.githubusercontent.com/154287/104357897-0e22d900-550e-11eb-9254-95d69cc938b6.png)

Feel free to move this issue to glossarist/concept-model if you find it more appropriate.

**Comment by skalee** [comment]
@ronaldtse ?

**Comment by skalee** [comment]
Also please note that there are concepts, for example in [713-10-41](http://www.electropedia.org/iev/iev.nsf/display?openform&ievref=713-10-41), in which some translation are marked as "qualifiers" whereas some other are regular parts of speech, like "adjective". Correct or anomaly?

![Zrzut ekranu 2021-01-14 o 11 16 17](https://user-images.githubusercontent.com/154287/104577897-4df6d700-565a-11eb-8ff9-10d6de1ef1ff.png)

**Comment by ronaldtse** [comment]
Sought clarification from IEC.

**Comment by skalee** [comment]
I think this is more about our internal data structures rather than about presentation layer.

**Comment by ronaldtse** [comment]
The internal data structure also depends on intention, so let's wait for IEC confirmation. It might take some time since some are on holiday.

**Comment by ronaldtse** [comment]
From IEC:

> This is yet another legacy problem.
>
> The rules were changed to say
> ![image](https://user-images.githubusercontent.com/11865/105430222-fece2900-5c8d-11eb-8c20-5d5965a21438.jpeg)
> ![image](https://user-images.githubusercontent.com/11865/105430238-05f53700-5c8e-11eb-83d6-ad05f501254f.jpeg)
>
> but TC 1 has yet to address all the legacy data.

Perhaps we have to mark this as legacy somehow. @skalee thoughts?

**Comment by ronaldtse** [comment]
To IEC:

In this case how should we deal with this to-be-retired attribute?

Perhaps in the model we keep it as a “legacy” free-text attribute, and in the user interface we show it with some warning (so people don’t enter this field in unless absolutely necessary?

ping @strogonoff for thoughts too.

**Comment by strogonoff** [comment]
@ronaldtse 

Background (which you may know): “qualifier” just means it’s a construct that _describes_ something (a thing or an action) and that can be removed from the sentence without messing up the grammar. These are some simple cases:

* _red wall_ — here “red“ is a qualifier;
* _AC voltage_ — here “AC” is a qualifier;
* _boiling water_ — here “boiling” is a qualifier.

Qualifiers are _often_ adjectives or adverbs, but other parts of speech could be used as qualifiers, especially if we are considering multi-word expressions (and vice-versa, adjectives may be used in a way that).

In my view, the “qualifier” flag isn’t entirely useless.  If I didn’t know it’s considered legacy, I’d be open to adding “qualifier” (or “modifier”, maybe even “premodifier” and “postmodifier”) marker in our [concept model](https://github.com/glossarist/glossarist-desktop/blob/master/src/models/concepts.ts#L181).

Pros:

- If the main word of an expression is a noun (“current”, “channel”), what is of more immediate importance to vocabulary reader: that it’s a “noun phrase”, or that it acts as a qualifier?
- Without the “qualifier” marker, there would be ambiguity as to what “part of speech” is supposed to mean. It’ll be tempting to mark a noun phrase as an “adjective” because it is used similarly to an adjective.

Cons:

- Presumably, the nature of being a “qualifier” could be conveyed through definition per the standard you quoted.
- The further one goes down this road, the more linguistic background would be required from the user to understand what’s going on (although we could dedicate more screen estate to dynamic help that would give some linguistics background to the user).
- A more precise and exhaustive model means more maintenance overhead and more possibilities for inconsistencies across concepts or languages within the same glossary.

**Comment by strogonoff** [comment]
If we are definitely retiring the qualifier, then we could do it the following way:

1) Where “qualifier” is given instead of part of speech, omit `partToSpeech` entirely (it’s optional).
2) Add “[qualifier]” as free text somewhere in definition (can be at the end or in the beginning). This way, the upcoming full-text search feature could be used to filter out all qualifiers for a batch update.

**Comment by ronaldtse** [comment]
@strogonoff IEC is going to remove all "qualifier" attributes sometime in the future, because the new rules no longer allow it. 

So keeping "qualifier" as an extra attribute in the model won't be useful in the future; we don't have anyone to use it, and the maintenance burden would be on us.

1. We will need to keep the "qualifier" for now because we need to faithfully reproduce IEC data.
2. We cannot "add" text to the definition because these data items are already standardized.

Could we handle this in some way that also allows handling other legacy attributes?

**Comment by strogonoff** [comment]
I see. 

We should probably switch to customizable schema for this, until then we have a couple options:

- [`usageInfo` freeform field](https://github.com/glossarist/glossarist-desktop/blob/4105c7a2b2b1f5085c748af3ce0fdb27fd7e3149/src/models/concepts.ts#L66). That applies to a concept as a whole; but in fact the trait of being “qualifier”

- The [`notes` field](https://github.com/glossarist/glossarist-desktop/blob/4105c7a2b2b1f5085c748af3ce0fdb27fd7e3149/src/models/concepts.ts#L76)

Those both apply to the concept as a whole, but technically the “qualifier” trait is one that should apply to the concept as opposed to particular designation🤔

**Comment by skalee** [comment]
Let's remove it now.

From data analysis, usage of this attribute is very inconsistent. For example these concepts have "qualifier" in term name:

- 151-15-01 pl, term name is "AC (kwalifikator)" (but no "qualifier" nor "kwalifikator" in term attributes)
- 131-14-23 pl, term name is "rozwarcia, kwalifikator"
- 151-15-13 pt, term name is "de entrada (qualificativo)"
- 161-05-01 it, term name is "ISM (qualificativo)"

And there are many, many more like these. I presume some other concepts have their definitions reworked to clearly indicate that it's qualifier.

I suppose the best we can do is to auto-replace "qualifier" termattribute with "(qualifier)" in designation or in front of definition. And perhaps to add some entry in `comments`.

----

FYI List of terms which have "qualifier" attribute or similar. May be incomplete if some localized term doesn't match my query or does not use latin script.

```
select ievref, language, term, termattribute from concepts where termattribute like '%qualif%';
```

produces:

<table>
<TR><TD>131-11-10</TD>
<TD>fr</TD>
<TD>à paramètres répartis</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>131-11-17</TD>
<TD>fr</TD>
<TD>indépendant du temps</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>131-11-19</TD>
<TD>fr</TD>
<TD>non linéaire</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>131-11-36</TD>
<TD>fr</TD>
<TD>non dissipatif</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>131-14-20</TD>
<TD>fr</TD>
<TD>en court-circuit</TD>
<TD>(1), qualificatif</TD>
</TR>
<TR><TD>131-14-21</TD>
<TD>en</TD>
<TD>short-circuit</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>131-14-21</TD>
<TD>fr</TD>
<TD>en court-circuit</TD>
<TD>(2), qualificatif</TD>
</TR>
<TR><TD>131-14-22</TD>
<TD>fr</TD>
<TD>en circuit ouvert</TD>
<TD>(1), qualificatif</TD>
</TR>
<TR><TD>131-14-23</TD>
<TD>en</TD>
<TD>open-circuit</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>131-14-23</TD>
<TD>fr</TD>
<TD>en circuit ouvert</TD>
<TD>(2), qualificatif</TD>
</TR>
<TR><TD>151-11-05</TD>
<TD>fr</TD>
<TD>d’électricité</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-11-05</TD>
<TD>pt</TD>
<TD>de electricidade</TD>
<TD>(qualificativo)</TD>
</TR>
<TR><TD>151-15-01</TD>
<TD>en</TD>
<TD>AC</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>151-15-01</TD>
<TD>fr</TD>
<TD>AC</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-02</TD>
<TD>en</TD>
<TD>DC</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>151-15-02</TD>
<TD>fr</TD>
<TD>DC</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-13</TD>
<TD>fr</TD>
<TD>d&#39;entrée</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-14</TD>
<TD>fr</TD>
<TD>de sortie</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-20</TD>
<TD>fr</TD>
<TD>en charge</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-21</TD>
<TD>fr</TD>
<TD>hors charge</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-58</TD>
<TD>fr</TD>
<TD>sous tension</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-15-59</TD>
<TD>fr</TD>
<TD>hors tension</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-16-05</TD>
<TD>fr</TD>
<TD>d&#39;extérieur</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>151-16-06</TD>
<TD>fr</TD>
<TD>d&#39;intérieur</TD>
<TD>qualificatif</TD>
</TR>
<TR><TD>161-05-01</TD>
<TD>en</TD>
<TD>ISM</TD>
<TD>(qualifier)</TD>
</TR>
<TR><TD>702-06-29</TD>
<TD>en</TD>
<TD>vestigial sideband</TD>
<TD>(qualifying term)</TD>
</TR>
<TR><TD>713-10-41</TD>
<TD>en</TD>
<TD>preselected-channel ...</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>826-14-15</TD>
<TD>en</TD>
<TD>inherently short-circuit and earth fault proof</TD>
<TD>qualifier</TD>
</TR>
<TR><TD>826-14-15</TD>
<TD>fr</TD>
<TD>intrinsèquement protégé contre les court-circuits et les défauts à la terre</TD>
<TD>qualificatif</TD>
</TR>
</table>

**Comment by strogonoff** [comment]
If we can remove it I don’t have objections. If we need to keep it for faithful data representation then I suggest to use an option from my previous comment (if it’s not a problem that “qualifier” will be applied to the whole concept and not to particular designation)

**Comment by ronaldtse** [comment]
@skalee we can't "remove" this attribute now, we must retain this because it is part of the standardized content.

> From data analysis, usage of this attribute is very inconsistent. For example, these concepts have "qualifier" in term name, not in term attributes

Qualifier is clearly an attribute, so we should extract that from the term/designation.

> And perhaps to add some entry in comments.

Maybe, or not. IEV manager says that the "qualifier" attribute (today) will transition to a sentence in "Notes" (in the future).

@strogonoff :
>  (if it’s not a problem that “qualifier” will be applied to the whole concept and not to particular designation)

In https://github.com/glossarist/iev-data/issues/94#issuecomment-760103051 @skalee provided an example where the "qualifier" attribute only applies to English.

@skalee are there any other instances that the "qualifier" attribute applies to a designation instead of a concept? If this is the only one I will ask IEC whether this particular one can be adjusted/fixed.

**Comment by strogonoff** [comment]
My reasoning suggests that, interestingly enough, being a qualifier is a property of the concept, not a particular designation 🤔

Unless we can come up with an example of a concept that has some designations that are qualifiers, while others are not (I couldn’t think of one).

(Whereas part of speech, abbreviation marker, etc. clearly belong to individual designations of that concept.)

**Comment by strogonoff** [comment]
(I.e., if we move the qualifier marker from designation to the concept, I believe that would be a lossless transformation.)

**Comment by strogonoff** [comment]
> My reasoning suggests that, interestingly enough, being a qualifier is a property of the concept, not a particular designation 🤔

I believe this logic is also supported by the document in @ronaldtse’s screenshot, which moves this aspect to the definition—which concerns the concept as a whole.

**Comment by ronaldtse** [comment]
I have asked IEC for confirmation on whether this attribute applies to the concept or designation.

Our original question:
>> Perhaps in the model we keep it as a “legacy” free-text attribute, and in the user interface we show it with some warning (so people don’t enter this field in unless absolutely necessary?

The IEV terminology manager has this to say regarding legacy attributes:
> from a general perspective, there will probably always be legacy issues since the standardization process is not immediate and so the proposal looks a good pragmatic approach from my perspective.

**Comment by skalee** [comment]
@ronaldtse:

> @skalee we can't "remove" this attribute now, we must retain this because it is part of the standardized content. Qualifier is clearly an attribute, so we should extract that from the term/designation.

To me it looks like they could be in the middle of transition from attribute to something else. Maybe we should ask for clarification.

@strogonoff:

> My reasoning suggests that, interestingly enough, being a qualifier is a property of the concept, not a particular designation 🤔

I have similar feeling.

By definition, term attributes are relevant to designation. They want to turn them into notes, what suggests that qualifiers are actually meant to be relevant to concepts, or at least localized concepts. This suggests misuse of term attributes.

@ronaldtse:

> @skalee are there any other instances that the "qualifier" attribute applies to a designation instead of a concept? If this is the only one I will ask IEC whether this particular one can be adjusted/fixed.

I'll investigate, though I'm not sure I'll find out anything — there is so much inconsistency.

**Comment by ronaldtse** [comment]
> To me it looks like they could be in the middle of transition from attribute to something else. Maybe we should ask for clarification.

They already said that they will be doing the transition but it might take years to revise these terms.

> I'll investigate, though I'm not sure I'll find out anything — there is so much inconsistency.

Thanks, please post if you find any inconsistency so that the IEV team can fix them (if editorial) or report them to the term owner for revision.

**Comment by skalee** [comment]
@ronaldtse This is really inconsistent… https://gist.github.com/skalee/e281031db37a1a08941f04ec1a7721af And I'm pretty sure I'll find even more when I improve the SQL query.

"AC" term is quite interesting — qualifier in Polish whereas abbreviation in Portuguese:

```
151-15-01        pl       AC (kwalifikator)
151-15-01        pt       CA (abreviatura)
```

**Comment by skalee** [comment]
@ronaldtse Got a question — are these IEV spreadsheets exports from some tool or they work on them directly? I'm asking because perhaps some discrepancies like "qualifier" being either in term column or in attribs column could be coming from export tool issue.

**Comment by ronaldtse** [comment]
@skalee these exports are exported from Lotus Domino. It reflects the actual data.

**Comment by ronaldtse** [comment]
Thanks for bringing this up, I will create a new issue on 151-15-01.

**Comment by ronaldtse** [comment]
From IEC:

> Strictly speaking the “qualifier" attribute applies at the concept level, but is inherited by the terms that are designated with a part of speech “adj” according to the revised rules. I don’t remember whether TBX allows this at the concept level, and I don’t have the time to look at present.
> 
> With time, the rule I provided the screen shots for below will be applied and thus the “qualifier” term attribute will disappear and be replaced by “adj”.
>
> FYI, as shown in the example in the IEC Supplement, SK.3.1.3.6.2 Grammatical information, the same term can represent a noun and an adjective:

![image001](https://user-images.githubusercontent.com/11865/106225276-c85d5480-621f-11eb-817f-21708b43244b.jpg)

tl;dr:
* "qualifier" attribute applies at concept level
* "qualifier" means each of its terms are considered to have a parts of speech "adj" according to the IEC revised rules (IEC Directives, Annex SK 3.1.3.6.2).
* in the future all "qualifier" attributes will be removed with these two changes:
  * "adj" applied to all its terms
  * the "qualifier" meaning will be expressed in the definition of its terms
* SK.3.1.3.6.2 specifically states that a term can be both a "noun" and "adj"

@skalee @strogonoff we will need to adjust the concept model to accommodate these. Thanks!

**Comment by ronaldtse** [comment]
Re: term can be both "noun" and "adj", I will seek clarification since the definitions can differ, and therefore the concept should also.

**Comment by strogonoff** [comment]
Requiring qualifiers to be strictly adjectives doesn’t make sense from linguistic point of view, since they aren’t always adjectives. A given vocabulary can impose that constraint, but I don’t think it generalizes across all vocabularies.

Furthermore, a designation can be (and often is) a multi-word expression, where each word belongs to its own part of speech. 

That’s why part of speech is an optional property of a designation in Glossarist’s model. A designation is either a qualifier or not, but it may not have a single clear part of speech, especially if it is a phrase consisting of multiple words.
